nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttp.java

```diff
@@ -0,0 +1,330 @@
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(ES_URL);
+        descriptors.add(PROP_SSL_CONTEXT_SERVICE);
+        descriptors.add(CONNECT_TIMEOUT);
+        descriptors.add(RESPONSE_TIMEOUT);
+        descriptors.add(ID_ATTRIBUTE);
+        descriptors.add(INDEX);
+        descriptors.add(TYPE);
+        descriptors.add(CHARSET);
+        descriptors.add(BATCH_SIZE);
+        descriptors.add(INDEX_OP);
+        return Collections.unmodifiableList(descriptors);
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> problems = new ArrayList<>(super.customValidate(validationContext));
+        // Since Expression Language is allowed for index operation, we can't guarantee that we can catch
+        // all invalid configurations, but we should catch them as soon as we can. For example, if the
+        // Identifier Attribute property is empty, the Index Operation must evaluate to "index".
+        String idAttribute = validationContext.getProperty(ID_ATTRIBUTE).getValue();
+        String indexOp = validationContext.getProperty(INDEX_OP).getValue();
+
+        if(StringUtils.isEmpty(idAttribute)) {
+            if(StringUtils.isEmpty(indexOp)) {
+                problems.add(new ValidationResult.Builder()
+                        .valid(false)
+                        .subject(INDEX_OP.getDisplayName())
+                        .explanation("If Identifier Attribute is not set, Index Operation must evaluate to \"index\"")
+                        .build());
+            }
+        }
+        return problems;
+    }
+
+    @OnScheduled
+    public void setup(ProcessContext context) {
+        super.setup(context);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final String id_attribute = context.getProperty(ID_ATTRIBUTE).getValue();
+        final Charset charset = Charset.forName(context.getProperty(CHARSET).getValue());
+
+        final List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+        OkHttpClient okHttpClient = getClient();
+        final ComponentLog logger = getLogger();
+
+        // Keep track of the list of flow files that need to be transferred. As they are transferred, remove them from the list.
+        List<FlowFile> flowFilesToTransfer = new LinkedList<>(flowFiles);
+        try {
+            final StringBuilder sb = new StringBuilder();
+            final String baseUrl = trimToEmpty(context.getProperty(ES_URL).getValue());
+            final URL url = new URL((baseUrl.endsWith("/") ? baseUrl : baseUrl + "/") + "_bulk");
+
+            for (FlowFile file : flowFiles) {
```

Comment:
What advantage do you get by doing batching in the onTrigger? It seems like each flowfile is sent in a single http request.